### PR TITLE
feat(templates): add javascript compiler

### PR DESCRIPTION
**Pull request by kmamal**
First iteration. I have done some basic testing and I think it works. Best way to try it is to run `yarn pack` in `nxt-lib` and then in the `asset` service `yarn install [path to pack file]`.

* Instead of using the `[[ ... ]]` syntax, I've decided to stick with `{{ ... }}`, and to extend it with a "language selector"
    * `{{ ... }}` uses old familiar nxtpressions
    * `{{#nxt ... }}` does the same
    * `{{#js ... }}` uses the new javascript syntax

* I've not used RecordAtom or mobx. There's some basic reactive logic inside an Observable. The only confusing part might be the `unusedRecordIds`? But let me know if you want me to do something else.
* I'm setting some of the node.js globals to `undefined` to avoid errors and including `_`, 'fp', and `moment` by default.

As an example the old nxtpression...

`{{ id | asset(\"general\") | ds(\":general.title\", \"value\") }}`

...will now look like...

`{{#js ds(ctx.id + ':asset.types')?.value?.includes('general') ? ds(ctx.id + ':general.title')?.value : undefined }}`

### CODE REVIEW

File: util/template/javascript.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+            let entry = activeRecords.get(recordId)
+            if (entry) {
+              return entry.value
+            }
+
+            entry = { value: undefined }
+            entry.subscription = ds.record
+              .observe(recordId)
+              .pipe(rx.distinctUntilChanged(fp.isEqual))
```
Comment (ronag) [inline util/template/javascript.js:59]:
```suggestion
```

**Review by ronag — COMMENTED** [review]
this needs tests

File: util/template/javascript.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                callScript()
+              })
+
+            activeRecords.set(recordId, entry)
+            return entry.value
+          }
+
+          let done = false
+
+          const callScript = _.debounce(() => {
```
Comment (ronag) [inline util/template/javascript.js:71]: i'd prefer just a next tick if "dirty"

**Comment by ronag** [comment]
I made a try for simplification:

```js
export default function ({ script, context, ds }) {
  return new rxjs.Observable(o => {
    const records = new Map()
    let dirty = false
    
    function callScript () {
      dirty = false
    
      try {
        o.next(script(context, getRecord)) // TODO (fix): https://nodejs.org/api/vm.html
      } catch (err) {
        o.error(err)
      }

      for (const [recordId, entry] of records) {
        entry.refs -= 1
        if (entry.refs === 0) {
          entry.record.unref()
          entry.record.off('update', refresh)
          records.delete(recordId)
        }
      }
    }
    
    function refresh () {
      if (!dirty) {
        dirty = true
        process.nextTick(callScript)
      }
    }
    
    function getRecord (recordId) {
      let entry = records.get(recordId)
      if (!entry) {
        entry = { refs: 0, record: ds.record.getRecord(recordId) }
        entry.record.on('update', refresh) // TODO (fix): Check for value change
        records.set(recordId, entry)
      }
      entry.refs += 1
      return entry.record.data
    }
    
    callScript()

    return () => {
      for (const entry of records) {
        entry.record.unref()
        entry.record.off('update', refresh)
      }
    }
  })
}
```

**Comment by kmamal** [comment]
Pushed fixes. Ref counting doesn't really work in this case because a record could be referenced twice in the same expression. Replaced it with a boolean `isActive`.

File: util/template/javascript.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            let entry = records.get(recordId)
+            if (!entry) {
+              entry = { isUsed: false, record: ds.record.getRecord(recordId) }
+              entry.record.on('update', refresh)
+              records.set(recordId, entry)
+            }
+            entry.isUsed = true
+
+            if (typeof state === 'string') {
+              // TODO (fix): Convert to state
```
Comment (ronag) [inline util/template/javascript.js:67]: Fix

File: util/template/javascript.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+const kEmpty = Symbol('kEmpty')
+
+module.exports = ({ ds } = {}) => {
+  return weakCache((expression) => {
+    try {
+      // eslint-disable-next-line no-new-func
+      const script = new Function(
+        ...overrides,
+        ...Object.keys(globals),
+        'ctx',
```
Comment (ronag) [inline util/template/javascript.js:39]: $

File: util/template/javascript.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        'ds',
```
Comment (ronag) [inline util/template/javascript.js:40]: nxt, nxt.ds

File: util/template/javascript.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  'process',
+  'queueMicrotask',
+  'require',
+  'setImmediate',
+  'setInterval',
+  'setTimeout',
+  'WebAssembly',
+]
+
+const globals = { _, fp, moment }
```
Comment (ronag) [inline util/template/javascript.js:28]: date-fns, luxor?

File: util/template/javascript.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+              // TODO (fix): Convert to state
+            }
+
+            if (entry.record.state < state) {
+              throw kEmpty
+            }
+
+            return entry.record.get(path)
+          }
+
```
Comment (ronag) [inline util/template/javascript.js:76]: const args = [                
  ...new Array(overrides.length),
   ...Object.values(globals),
   context,
   getRecord
]

File: util/template/javascript.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  'fetch',
+  'global',
+  'module',
+  'process',
+  'queueMicrotask',
+  'require',
+  'setImmediate',
+  'setInterval',
+  'setTimeout',
+  'WebAssembly',
```
Comment (ronag) [inline util/template/javascript.js:25]: ...Object.keys(globalThis)

Comment (ronag) [reply util/template/javascript.js:25]: https://nodejs.org/api/globals.html

File: util/template/javascript.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+              entry = { isUsed: false, record: ds.record.getRecord(recordId) }
+              entry.record.on('update', refresh)
+              records.set(recordId, entry)
+            }
+            entry.isUsed = true
+
+            if (typeof state === 'string') {
+              // TODO (fix): Convert to state
+            }
+
```
Comment (ronag) [inline util/template/javascript.js:69]:
```suggestion
if (state == null) {
  state = recordId.startsWith('{') || recordId.includes('?') ? ds.record.PROVIDER  ds.record.SERVER
}
```

File: util/template/javascript.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+const globals = { _, fp, moment }
+
+const kEmpty = Symbol('kEmpty')
+
+module.exports = ({ ds } = {}) => {
+  return weakCache((expression) => {
+    try {
+      // eslint-disable-next-line no-new-func
+      const script = new Function(
+        ...overrides,
```
Comment (deadbeef84) [inline util/template/javascript.js:37]: Perhaps spread overrides at end, then you wont even have to pass them when calling the function.

Comment (deadbeef84) [reply util/template/javascript.js:37]: or including overrides in globals, and let them be set to undefined there

File: util/template/javascript.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+          const callScript = () => {
+            dirty = false
+
+            for (const entry of records.values()) {
+              entry.isUsed = false
+            }
+
+            try {
+              // TODO (fix): https://nodejs.org/api/vm.html
```
Comment (deadbeef84) [inline util/template/javascript.js:85]: Using the vm module would be nice, then you wouldn't need `ctx.x` but could access context variables directly... I guess you could do the same here, but then you would need to recreate the function whenever context keys change.

Comment (deadbeef84) [reply util/template/javascript.js:85]: I guess vm module also complicates the getRecord method...?

Comment (ronag) [reply util/template/javascript.js:85]: I think the problem with vm is passing stuff in and out... I'm not sure it will work

**Comment by ronag** [comment]
```js
nxt = { 
  ds, // se old expression code
  hashaint, // se old expression code
  asset, // se old expression code
  timer(dueTime, undueValue, dueValue) // re-evaluate after due time and return corresponding value
}
```

**Comment by kmamal** [comment]
Pushed fixes.
* Using node vm for execution. The global scope is empty in vm which is nice.
* I'm now puttin the context variables (now called `args` to differentiate them from the vm context) in the global scope.
* pipeline operator is in, using hack pipes and the # placeholder
* ds is inside `nxt` and so are `asset` and `hashaint`.

@ronag I'm not exactly sure I understand the timer function. I've put a TODO.

File: util/template/javascript.js [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+const vm = require('node:vm')
+const babel = require('@babel/core')
+
+const babelOptions = {
+  plugins: [
+    [
+      '@babel/plugin-proposal-pipeline-operator',
+      {
+        proposal: 'hack',
+        topicToken: '#',
```
Comment (ronag) [inline util/template/javascript.js:14]: I think % is more likely for final proposal?

File: util/template/javascript.js [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+            if (entry.record.state < state) {
+              throw kEmpty
+            }
+
+            return entry.record.get(path)
+          }
+
+          const hasAssetType = (id, type, _return = true) => {
+            const types = getRecord(id + ':asset.rawTypes?', 'value')
+            return types.includes(type)
```
Comment (ronag) [inline util/template/javascript.js:76]: fp.includes is more defensive

File: util/template/javascript.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            }
+
+            if (entry.record.state < state) {
+              throw kEmpty
+            }
+
+            return entry.record.get(path)
+          }
+
+          const hasAssetType = (id, type, _return = true) => {
```
Comment (ronag) [inline util/template/javascript.js:74]: _return does nothing now, maybe throw kEmpty?

**Review by ronag — CHANGES_REQUESTED** [review]

File: util/template/javascript.js (hunk 13 above)
Comment (ronag) [reply util/template/javascript.js:76]: Or at least ?.includes

File: util/template/javascript.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+          let dirty = false
+
+          const refresh = () => {
+            if (!dirty) {
+              dirty = true
+              setImmediate(callScript)
+            }
+          }
+
+          const getRecord = (recordId, path, state = ds.record.SERVER) => {
```
Comment (ronag) [inline util/template/javascript.js:47]: Remove default arg for state

File: util/template/javascript.js (hunk 13 above)
Comment (ronag) [inline util/template/javascript.js:76]: Return id if true

**Comment by ronag** [comment]
```
commit fd98541e08a590ce9ca1df3983f3d430213313f0
Author: Robert Nagy <ronagy@icloud.com>
Date:   Tue Oct 11 20:50:59 2022 +0200

    fixup

diff --git a/util/template/javascript.js b/util/template/javascript.js
index c5d796c..09ff5d6 100644
--- a/util/template/javascript.js
+++ b/util/template/javascript.js
@@ -23,7 +23,14 @@ const globals = {
   moment: require('moment'),
 }
 
-const kEmpty = Symbol('kEmpty')
+const kWait = Symbol('kWait')
+const kNull = Symbol('kNull')
+
+const hashaint = (value, _options = {}) => {
+  const str = JSON.stringify(value)
+  const hash = hasha(str, _options)
+  return parseInt(hash.slice(-13), 16)
+}
 
 module.exports = ({ ds } = {}) => {
   return weakCache((expression) => {
@@ -59,27 +66,27 @@ module.exports = ({ ds } = {}) => {
 
             if (state == null) {
               state =
-                recordId.startsWith('{') || recordId.endsWith('?')
+                recordId.startsWith('{') || recordId.includes('?')
                   ? ds.record.PROVIDER
                   : ds.record.SERVER
             }
 
             if (entry.record.state < state) {
-              throw kEmpty
+              throw kWait
             }
 
             return entry.record.get(path)
           }
 
-          const hasAssetType = (id, type, _return = true) => {
+          const hasAssetType = (id, type, throws = true) => {
             const types = getRecord(id + ':asset.rawTypes?', 'value')
-            return types.includes(type)
-          }
-
-          const hashaint = (value, _options = {}) => {
-            const str = JSON.stringify(value)
-            const hash = hasha(str, _options)
-            return parseInt(hash.slice(-13), 16)
+            if (types.includes(type)) {
+              return id
+            }
+            if (throws) {
+              throw kNull
+            }
+            return null
           }
 
           const context = vm.createContext({
@@ -103,7 +110,9 @@ module.exports = ({ ds } = {}) => {
             try {
               o.next(script.runInContext(context))
             } catch (err) {
-              if (err !== kEmpty) {
+              if (err === kNull) {
+                o.next(null)
+              } else if (err !== kWait) {
                 o.error(err)
               }
             }

```

**Comment by ronag** [comment]
```patch
commit 6e97b6f753dccf9127daf7e00b81dd5b27086dad
Author: Robert Nagy <ronagy@icloud.com>
Date:   Tue Oct 11 20:50:59 2022 +0200

    fixup

diff --git a/util/template/javascript.js b/util/template/javascript.js
index c5d796c..aedf852 100644
--- a/util/template/javascript.js
+++ b/util/template/javascript.js
@@ -23,7 +23,14 @@ const globals = {
   moment: require('moment'),
 }
 
-const kEmpty = Symbol('kEmpty')
+const kWait = Symbol('kWait')
+const kNull = Symbol('kNull')
+
+const hashaint = (value, _options = {}) => {
+  const str = JSON.stringify(value)
+  const hash = hasha(str, _options)
+  return parseInt(hash.slice(-13), 16)
+}
 
 module.exports = ({ ds } = {}) => {
   return weakCache((expression) => {
@@ -44,7 +51,7 @@ module.exports = ({ ds } = {}) => {
             }
           }
 
-          const getRecord = (recordId, path, state = ds.record.SERVER) => {
+          const getRecord = (recordId, path, state) => {
             let entry = records.get(recordId)
             if (!entry) {
               entry = { isUsed: false, record: ds.record.getRecord(recordId) }
@@ -59,27 +66,27 @@ module.exports = ({ ds } = {}) => {
 
             if (state == null) {
               state =
-                recordId.startsWith('{') || recordId.endsWith('?')
+                recordId.startsWith('{') || recordId.includes('?')
                   ? ds.record.PROVIDER
                   : ds.record.SERVER
             }
 
             if (entry.record.state < state) {
-              throw kEmpty
+              throw kWait
             }
 
             return entry.record.get(path)
           }
 
-          const hasAssetType = (id, type, _return = true) => {
-            const types = getRecord(id + ':asset.rawTypes?', 'value')
-            return types.includes(type)
-          }
-
-          const hashaint = (value, _options = {}) => {
-            const str = JSON.stringify(value)
-            const hash = hasha(str, _options)
-            return parseInt(hash.slice(-13), 16)
+          const hasAssetType = (id, type, throws = true) => {
+            const types = getRecord(id + ':asset.rawTypes?', 'value', ds.record.PROVIDER)
+            if (types.includes(type)) {
+              return id
+            }
+            if (throws) {
+              throw kNull
+            }
+            return null
           }
 
           const context = vm.createContext({
@@ -103,7 +110,9 @@ module.exports = ({ ds } = {}) => {
             try {
               o.next(script.runInContext(context))
             } catch (err) {
-              if (err !== kEmpty) {
+              if (err === kNull) {
+                o.next(null)
+              } else if (err !== kWait) {
                 o.error(err)
               }
             }

```

**Comment by ronag** [comment]
I think all nxt context can live under `nxt.`, i.e. `nxt.id`

**Comment by ronag** [comment]
https://github.com/babel/babel/issues/15039#issuecomment-1275629199

**Comment by ronag** [comment]
@deadbeef84 I have already done that: https://github.com/nxtedition/nxt/pull/8574

**Comment by ronag** [comment]
What I'm missing from e.g. fp.pipe is the ability to short-circuit (which to be fair we don't have in hack pipes yet either).

**Comment by deadbeef84** [comment]
> What I'm missing from e.g. fp.pipe is the ability to short-circuit (which to be fair we don't have in hack pipes yet either).

yeah it's mentioned as a future extension for hack pipes, so it's not part of the current proposal so probably wont see it anytime soon. I think our closest bet is something similar to `return()` in nxtpressions? Assuming you want to short-circuit the entire expression :grimacing:

File: util/template/javascript.js [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+            return entry.record.get(path)
+          }
+
+          const hasAssetType = (id, type, throws = true) => {
+            const types = getRecord(id + ':asset.rawTypes?', 'value')
+            if (types.includes(type)) {
+              return id
+            }
+            if (throws) {
```
Comment (ronag) [inline util/template/javascript.js:78]: I don't think we need this anymore...

**Review by ronag — APPROVED** [review]
LGTM, is this ready?

**Comment by kmamal** [comment]
* removed hack pipes
* added `nxt.pipe`
* I have made a new PR for changes expressions: https://github.com/nxtedition/nxt/pull/8580

**Comment by kmamal** [comment]
I'd say it's ready yes.

File: util/template/javascript.js [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+              value = func(value)
+              if (value == null) {
+                break
+              }
+            }
+            return value
+          }
+
+          pipe.asset = (type) => (id) => hasAssetType(id, type, false)
+          pipe.ds = (domain, path, state) => (id) => getRecord(id + domain, path, state)
```
Comment (ronag) [inline util/template/javascript.js:126]: Maybe we should skip these to begin with?

File: util/template/index.js [hunk 18]
```diff
@@ -1,12 +1,16 @@
 const rx = require('rxjs/operators')
 const Observable = require('rxjs')
 const fp = require('lodash/fp')
-const getExpressionCompiler = require('./expression')
+const getNxtpressionsCompiler = require('./nextpressions')
```
Comment (deadbeef84) [inline util/template/index.js:4]:
```suggestion
const getNxtpressionsCompiler = require('./nxtpressions')
```

File: util/template/javascript.js [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+
+          const context = vm.createContext({
+            ...globals,
+            $: args,
+            nxt: {
+              ...args,
```
Comment (deadbeef84) [inline util/template/javascript.js:125]: remove args here?

**Pull request merged**